[ML] Make datafeeds work with nanosecond time fields

droberts195 opened the pull request:
Allows ML datafeeds to work with time fields that have
the "date_nanos" type _and make use of the extra precision_.
(Previously datafeeds only worked with time fields that were
exact multiples of milliseconds.  So datafeeds would work
with "date_nanos" only if the extra precision over "date" was
not used.)

Relates #49889

droberts195 committed:
[ML] Make datafeeds work with nanosecond time fields

Allows ML datafeeds to work with time fields that have
the "date_nanos" type.  (Previously datafeeds only
worked with time fields that had the "date" type.)

elasticmachine commented:
Pinging @elastic/ml-core (:ml)

davidkyle commented on x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/extractor/TimeFieldTests.java:
```diff
@@ -5,20 +5,48 @@
+    public void testDocValueWithFractionalMillisecondStringValue() {
+        long millis = randomNonNegativeLong();
+        int extraNanos = randomIntBetween(0, 999999);
+        Instant time = Instant.ofEpochSecond(millis / 1000, (millis % 1000) * 1000000 + extraNanos);
```
```suggestion
        Instant time = Instant.ofEpochMilli(millis).plusNanos(extraNanos);
```

davidkyle reviewed (approved):
LGTM

droberts195 committed:
Address review comment

Co-Authored-By: David Kyle <david.kyle@elastic.co>

droberts195 committed:
Merge branch 'master' into make_datafeed_work_with_nanosecond_time

droberts195 committed:
Detect "date_nanos" time fields through more code paths

The extractor for the primary time field of an anomaly
detection job is automatically created as a TimeField
without checking its field caps response, so earlier
changes in this PR did enough to make that scenario
work for "date_nanos" fields.  However, the main field
type detection code was still only accepting "date"
to create a TimeField.  This change makes it accept
either or both of "date" and "date_nanos".

droberts195 committed:
Merge branch 'make_datafeed_work_with_nanosecond_time' of github.com:droberts195/elasticsearch into make_datafeed_work_with_nanosecond_time

droberts195 committed:
Use low level functions rather than regex to remove fraction

Profiling showed the low level version is 3 times faster
and we do time parsing for every single document fed to
a job.

przemekwitek reviewed (approved):
LGTM

dimitris-athanasiou reviewed (approved):
LGTM2
